src/deviceconfigurator.cpp

```diff
@@ -2006,55 +2006,53 @@ QString DeviceTreeXmlHandler::getXmlData(const QString &machine)
+	if ( xmlMachineEntry.readNextStartElement() ) {
+		if ( xmlMachineEntry.name() == "machine" ) {
+			while ( xmlMachineEntry.readNextStartElement() ) {
+				if ( xmlMachineEntry.name() == "description" ) {
+					QString description(xmlMachineEntry.readElementText());
+					QTextDocument doc;
+					doc.setHtml(description);
+					return doc.toPlainText();
+				} else
+					xmlMachineEntry.skipCurrentElement();
+			}
+		}
 	} else
 		return QString();
 }
 
 QString DeviceTreeXmlHandler::lookupBiosOptions(const QString &machine, QStringList *bioses, QStringList *biosDescriptions)
 {
-	QStringList xmlLines(qmc2MachineList->xmlDb()->xml(machine).split('\n'));
-	if ( xmlLines.count() > 1 ) {
-		int index = 0;
-		QString defaultOption;
-		while ( index < xmlLines.count() ) {
-			if ( xmlLines.at(index).startsWith("<biosset ") ) {
-				QString name;
-				QString description;
-				int startIndex = xmlLines.at(index).indexOf("name=\"");
-				if ( startIndex >= 0 ) {
-					startIndex += 6;
-					int endIndex = xmlLines.at(index).indexOf("\"", startIndex);
-					name = xmlLines.at(index).mid(startIndex, endIndex - startIndex);
-				}
-				startIndex = xmlLines.at(index).indexOf("description=\"");
-				if ( startIndex >= 0 ) {
-					startIndex += 13;
-					int endIndex = xmlLines.at(index).indexOf("\"", startIndex);
-					description = xmlLines.at(index).mid(startIndex, endIndex - startIndex);
-					QTextDocument doc;
-					doc.setHtml(description);
-					description = doc.toPlainText();
-				}
-				if ( !name.isEmpty() && !description.isEmpty() ) {
-					bioses->append(name);
-					biosDescriptions->append(description);
-					if ( xmlLines.at(index).indexOf("default=\"yes\"") >= 0 )
-						defaultOption = name;
-				}
+	QXmlStreamReader xmlMachineEntry(qmc2MachineList->xmlDb()->xml(machine));
+	if ( xmlMachineEntry.readNextStartElement() ) {
+		if ( xmlMachineEntry.name() == "machine" ) {
+			QString defaultOption;
+			while ( xmlMachineEntry.readNextStartElement() ) {
+				if ( xmlMachineEntry.name() == "biosset" ) {
+					QString name;
+					QString description;
+					if ( xmlMachineEntry.attributes().hasAttribute("name") )
+						name = xmlMachineEntry.attributes().value("name").toString();
+					if ( xmlMachineEntry.attributes().hasAttribute("description") ) {
+						description = xmlMachineEntry.attributes().value("description").toString();
+						QTextDocument doc;
+						doc.setHtml(description);
+						description = doc.toPlainText();
+					}
+					if ( !name.isEmpty() && !description.isEmpty() ) {
+						bioses->append(name);
+						biosDescriptions->append(description);
+						if ( xmlMachineEntry.attributes().hasAttribute("default") && xmlMachineEntry.attributes().value("default").toString() == "yes" )
+							defaultOption = name;
+					}
+					xmlMachineEntry.skipCurrentElement();
+				} else
+					xmlMachineEntry.skipCurrentElement();
 			}
-			index++;
+			return defaultOption;
 		}
-		return defaultOption;
 	} else
 		return QString();
```

Comment:
This now has a “not all paths return a value” issue.  The compiler should be able to catch it – is that warning disabled?  In the case where the root element isn’t a `machine` element (which shouldn’t happen unless the user hacks the database manually, it will fall off the end without returning a value.

You can change the end of it to always return a value:
```c++
				} else
					xmlMachineEntry.skipCurrentElement();
			}
			index++;
			return defaultOption;
		}
		return defaultOption;
	}
	return QString();
```

Reply:
This should be fixed now.